### PR TITLE
[SYCL-MLIR] Promote f16 operands of unary arithmetic expressions

### PR DESCRIPTION
**Pull request by victor-eds**
Also, the `__imag__` operator now returns 0 when used on a scalar.

### CODE REVIEW

File: polygeist/tools/cgeist/Test/Verification/float16.c [hunk 1]
```diff
@@ -23,9 +23,9 @@ _Float16 type(_Float16 arg) {
 // CHECK-EXTEND-NEXT:     %[[EXT1:.*]] = arith.extf %arg1 : f16 to f32
 // CHECK-EXTEND-NEXT:     %[[ADD:.*]] = arith.addf %[[EXT0]], %[[EXT1]] : f32
 // CHECK-EXTEND-NEXT:     %[[EXT2:.*]] = arith.extf %arg2 : f16 to f32
-// CHECK-EXTEND-NEXT:     %[[NEG:.*]] = arith.negf %arg3 : f16
-// CHECK-EXTEND-NEXT:     %[[EXTNEG:.*]] = arith.extf %[[NEG]] : f16 to f32
-// CHECK-EXTEND-NEXT:     %[[MUL:.*]] = arith.mulf %[[EXT2]], %[[EXTNEG]] : f32
+// CHECK-EXTEND-NEXT:     %[[EXT3:.*]] = arith.extf %arg3 : f16 to f32
+// CHECK-EXTEND-NEXT:     %[[NEG:.*]] = arith.negf %[[EXT3]] : f32
+// CHECK-EXTEND-NEXT:     %[[MUL:.*]] = arith.mulf %[[EXT2]], %[[NEG]] : f32
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Test/Verification/float16.c:28]: We promote now

File: polygeist/tools/cgeist/Test/Verification/imag.c [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+// CHECK-LABEL:   func.func @f1(
+// CHECK-SAME:                  %[[VAL_0:.*]]: f32) -> f32 attributes {llvm.linkage = #llvm.linkage<external>} {
+// CHECK:           %[[VAL_1:.*]] = arith.constant 0.000000e+00 : f32
+// CHECK:           return %[[VAL_1]] : f32
+// CHECK:         }
+
+float f1(float a) {
+  return __imag__(a);
+}
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Test/Verification/imag.c]: This was not working

File: polygeist/tools/cgeist/Test/Verification/imag_complex.c [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+#include <complex.h>
+
+int f2(int complex a) {
+  return __imag__(a);
+}
+
+float f3(float complex a) {
+  return __imag__(a);
+}
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Test/Verification/imag_complex.c:12]: complex not working at all, so this is out of scope.

File: polygeist/tools/cgeist/Lib/clang-mlir.h [hunk 4]
```diff
@@ -521,7 +521,13 @@ class MLIRScanner : public clang::StmtVisitor<MLIRScanner, ValueCategory> {
...
   ValueCategory VisitBin##OP##Assign(clang::BinaryOperator *E);
 #include "Expressions.def"
-#undef HANDLEBINOP
+
+#define HANDLEUNARYOP(OP)                                                      \
+  ValueCategory VisitUnary##OP(clang::UnaryOperator *E,                        \
+                               clang::QualType PromotionTy =                   \
+                                   clang::QualType());                         \
+  ValueCategory Visit##OP(clang::UnaryOperator *E, clang::QualType PromotionTy);
+#include "Expressions.def"
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/clang-mlir.h:530]: After macro has been used we should #undef it.

File: polygeist/tools/cgeist/Lib/ValueCategory.cc [hunk 5]
```diff
@@ -54,6 +55,31 @@ mlir::Value ValueCategory::getValue(mlir::OpBuilder &builder) const {
   llvm_unreachable("type must be LLVMPointer or MemRef");
 }
 
+ValueCategory ValueCategory::getNullValue(OpBuilder &Builder, Location Loc,
+                                          Type Type) {
+  const auto Value =
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/ValueCategory.cc:60]: [nit]: Value -> ZeroVal

File: polygeist/tools/cgeist/Lib/ValueCategory.cc [hunk 6]
```diff
@@ -632,3 +658,15 @@ ValueCategory ValueCategory::InBoundsGEPOrSubIndex(OpBuilder &Builder,
...
+
+template <typename OpTy>
+ValueCategory FPUnaryOp(OpBuilder &Builder, Location Loc, Value Val) {
+  assert(mlirclang::isFPOrFPVectorTy(Val.getType()) &&
+         "Expecting FP or FP vector operand type");
+  warnUnconstrainedOp<arith::NegFOp>();
+  return {Builder.createOrFold<OpTy>(Loc, Val), false};
+}
+
+ValueCategory ValueCategory::FNeg(OpBuilder &Builder, Location Loc) const {
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/ValueCategory.cc:670]: Can we pass the value to negate in the parameter list (like it is passed for FPUnaryOp) pls.

File: polygeist/tools/cgeist/Lib/ValueCategory.h [hunk 7]
```diff
@@ -99,6 +102,7 @@ class ValueCategory {
                      mlir::Value RHS, bool IsExact = false) const;
   ValueCategory ExactSDiv(mlir::OpBuilder &Builder, mlir::Location Loc,
                           mlir::Value RHS) const;
+  ValueCategory FNeg(mlir::OpBuilder &Builder, mlir::Location Loc) const;
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/ValueCategory.h:105]: I think we should pass the value to be negated in the parameter list.

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 8]
```diff
@@ -2901,3 +2903,76 @@ ValueCategory MLIRScanner::EmitBinOr(const BinOpInfo &Info) {
...
+            ? Glob.getTypes().getPromotionType(E->getSubExpr()->getType())     \
+            : PromotionType;                                                   \
+    auto result = Visit##OP(E, promotionTy);                                   \
+    if (result.val && !promotionTy.isNull())                                   \
+      result = EmitUnPromotedValue(getMLIRLocation(E->getExprLoc()), result,   \
+                                   E->getType());                              \
+    return result;                                                             \
+  }
+#include "Expressions.def"
+
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:2926]: Add #undef HANDLEUNARYOP

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 9]
```diff
@@ -2901,3 +2903,76 @@ ValueCategory MLIRScanner::EmitBinOr(const BinOpInfo &Info) {
...
+    LLVM_DEBUG({                                                               \
+      llvm::dbgs() << "VisitUnary" #OP ": ";                                   \
+      E->dump();                                                               \
+      llvm::dbgs() << "\n";                                                    \
+    });                                                                        \
+    QualType promotionTy =                                                     \
+        PromotionType.isNull()                                                 \
+            ? Glob.getTypes().getPromotionType(E->getSubExpr()->getType())     \
+            : PromotionType;                                                   \
+    auto result = Visit##OP(E, promotionTy);                                   \
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:2919]: Use actual type returned by the Visit member function here.

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 10]
```diff
@@ -2901,3 +2903,76 @@ ValueCategory MLIRScanner::EmitBinOr(const BinOpInfo &Info) {
...
+
+ValueCategory MLIRScanner::VisitPlus(UnaryOperator *E, QualType PromotionType) {
+  if (!PromotionType.isNull())
+    return EmitPromotedScalarExpr(E->getSubExpr(), PromotionType);
+  return Visit(E->getSubExpr());
+}
+
+ValueCategory MLIRScanner::VisitMinus(UnaryOperator *E,
+                                      QualType PromotionType) {
+  const auto Loc = getMLIRLocation(E->getExprLoc());
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:2935]: auto -> Location

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 11]
```diff
@@ -2901,3 +2903,76 @@ ValueCategory MLIRScanner::EmitBinOr(const BinOpInfo &Info) {
...
+    Op = EmitPromotedScalarExpr(E->getSubExpr(), PromotionType);
+  else
+    Op = Visit(E->getSubExpr());
+
+  // Generate a unary FNeg for FP ops.
+  if (mlirclang::isFPOrFPVectorTy(Op.val.getType()))
+    return Op.FNeg(builder, Loc);
+
+  // Emit unary minus with EmitBinSub so we handle overflow cases etc.
+  const auto Zero = ValueCategory::getNullValue(builder, Loc, Op.val.getType());
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:2947]: auto -> ValueCategory

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 12]
```diff
@@ -2901,3 +2903,76 @@ ValueCategory MLIRScanner::EmitBinOr(const BinOpInfo &Info) {
...
+  return EmitBinSub(
+      BinOpInfo{Zero, Op, E->getType(), BinaryOperator::Opcode::BO_Sub, E});
+}
+
+ValueCategory MLIRScanner::VisitImag(UnaryOperator *E, QualType PromotionType) {
+  Expr *Op = E->getSubExpr();
+
+  assert(!Op->getType()->isAnyComplexType() && "Unsupported");
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:2955]: A bit confusing, the member function name `VisitImag` made me think we would receive an expression that has Comples type and just visit the immaginary subexpression. Can you explain in a comment perhaps the intention of this function?

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 13]
```diff
@@ -2901,3 +2903,76 @@ ValueCategory MLIRScanner::EmitBinOr(const BinOpInfo &Info) {
...
+    Visit(Op);
+  auto ResTy = Glob.getTypes().getMLIRType(
+      !PromotionType.isNull() ? PromotionType : E->getType());
+  return ValueCategory::getNullValue(builder, loc, ResTy);
+}
+
+ValueCategory MLIRScanner::VisitReal(UnaryOperator *E, QualType PromotionType) {
+  Expr *Op = E->getSubExpr();
+
+  assert(!Op->getType()->isAnyComplexType() && "Unsupported");
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:2973]: Same observation as for line 2955

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 14]
```diff
@@ -2129,21 +2129,21 @@ static bool isSigned(QualType Ty) {
...
-  constexpr const BinaryOperator *getExpr() const { return Expr; }
+  constexpr const Expr *getExpr() const { return E; }
 
 private:
   const ValueCategory LHS;
   const ValueCategory RHS;
   const QualType Ty;                   // Computation Type.
   const BinaryOperator::Opcode Opcode; // Opcode of BinOp to perform
-  const BinaryOperator *Expr;
+  const Expr *E;
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:2146]: I see the type of `E` is now a generic `Expr`. What was the motivation for not using `BinaryOperator`?

File: polygeist/tools/cgeist/Test/Verification/imag_complex.c (hunk 3 above)
Comment (etiotto) [reply polygeist/tools/cgeist/Test/Verification/imag_complex.c:12]: I'd would have wanted this test case to be merged with imag.c  but I see you need to XFAIL it for now (so it has to be a separate test).

File: polygeist/tools/cgeist/Lib/clang-mlir.h (hunk 4 above)
Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/clang-mlir.h:530]: It's being `#undef`ed in the `Expressions.def` file.

File: polygeist/tools/cgeist/Lib/ValueCategory.cc (hunk 6 above)
Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/ValueCategory.cc:670]: This is a member function. Following the convention of the rest of the functions, we do not pass the "first" argument.

File: polygeist/tools/cgeist/Lib/ValueCategory.h (hunk 7 above)
Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/ValueCategory.h:105]: See above

File: polygeist/tools/cgeist/Lib/CGExpr.cc (hunk 8 above)
Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:2926]: Answered above

File: polygeist/tools/cgeist/Lib/CGExpr.cc (hunk 12 above)
Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:2955]: We can receive any scalar expression AFAIC here. `__imag__(X) = 0` for every scalar `X`.

Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:2955]: There is already a comment below explaining the behavior for scalar input expressions.

File: polygeist/tools/cgeist/Lib/CGExpr.cc (hunk 13 above)
Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:2973]: See response above.

File: polygeist/tools/cgeist/Lib/CGExpr.cc (hunk 14 above)
Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:2146]: To be able to reuse this to perform `-X == 0 - X` (for integers).

File: polygeist/tools/cgeist/Lib/clang-mlir.h (hunk 4 above)
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/clang-mlir.h:530]: Ok all good then.

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 15]
```diff
@@ -2554,7 +2555,6 @@ ValueCategory MLIRScanner::EmitCompoundAssign(
                               &MLIRScanner::EmitBin##OP);                      \
   }
 #include "Expressions.def"
-#undef HANDLEBINOP
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:2557]: I'd prefer the undef to be outside the definition file.

File: polygeist/tools/cgeist/Lib/CGExpr.cc (hunk 8 above)
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:2890]: I'd prefer the undef to be outside the definition file. Clearer IMO.

**Review by etiotto — APPROVED** [review]
I have some minor comments pending, but overall, this PR is good to go. Thanks Victor!

**Review by whitneywhtsang — APPROVED** [review]
